components/script/dom/webgltransformfeedback.rs

```diff
@@ -0,0 +1,126 @@
+use crate::dom::bindings::codegen::Bindings::WebGLTransformFeedbackBinding;
+use crate::dom::bindings::reflector::{reflect_dom_object, DomObject};
+use crate::dom::bindings::inheritance::Castable;
+use crate::dom::bindings::root::DomRoot;
+use crate::dom::webglobject::WebGLObject;
+use crate::dom::webglrenderingcontext::WebGLRenderingContext;
+use canvas_traits::webgl::{webgl_channel, WebGLCommand};
+use dom_struct::dom_struct;
+use std::cell::Cell;
+
+#[dom_struct]
+pub struct WebGLTransformFeedback {
+    webgl_object: WebGLObject,
+    id: u32,
+    marked_for_deletion: Cell<bool>,
+    has_been_bound: Cell<bool>,
+    is_active: Cell<bool>,
+    is_paused: Cell<bool>,
+}
+
+impl WebGLTransformFeedback {
+    fn new_inherited(context: &WebGLRenderingContext, id: u32) -> Self {
+        Self {
+            webgl_object: WebGLObject::new_inherited(context),
+            id,
+            marked_for_deletion: Cell::new(false),
+            has_been_bound: Cell::new(false),
+            is_active: Cell::new(false),
+            is_paused: Cell::new(false),
+        }
+    }
+
+    pub fn new(context: &WebGLRenderingContext) -> DomRoot<Self> {
+        let (sender, receiver) = webgl_channel().unwrap();
+        context.send_command(WebGLCommand::CreateTransformFeedback(sender));
+        let id = receiver.recv().unwrap();
+
+        reflect_dom_object(
+            Box::new(WebGLTransformFeedback::new_inherited(context, id)),
+            &*context.global(),
+            WebGLTransformFeedbackBinding::Wrap,
+        )
+    }
+}
+
+impl WebGLTransformFeedback {
+    pub fn bind(&self, context: &WebGLRenderingContext, target: u32) {
+        context.send_command(WebGLCommand::BindTransformFeedback(target, self.id()));
+        self.has_been_bound.set(true);
+    }
+
+    pub fn begin(&self, context: &WebGLRenderingContext, primitive_mode: u32) {
+        if self.has_been_bound.get() && !self.is_active() {
+            context.send_command(WebGLCommand::BeginTransformFeedback(primitive_mode));
+            self.set_active(true);
+        }
+    }
+
+    pub fn end(&self, context: &WebGLRenderingContext) {
+        if self.has_been_bound.get() && self.is_active() {
```

Comment:
According to the GLES spec, we should implicitly resume if we're paused when ending a transform feedback.